### DISCUSSION TITLE
Questions about KTX2 in v8

### DISCUSSION THREAD

**Discussion by Sparcks** [pixijs/pixijs#10222] [v8 General] [answered]
Hi

I'm trying out ktx2 in v8 (rc9) and stumbled upon an issue. I created several .ktx2 files (etc1s & uastc, all of them 777x1198) with different settings and none of them are loading, only the .png is loading. I also created a sprite animation with multiple .ktx2 files (one is etc1s and one uastc, all of them 1920x1080) starting from a .jpg and these load without an issue.

_NOTE: Don't mind the 'KTX2 Disabled' button, that's for a different test._
![Screenshot 2024-02-16 110838](https://github.com/pixijs/pixijs/assets/4526795/56f0f578-6484-4374-b202-cca710a16e55)
![Screenshot 2024-02-16 110906](https://github.com/pixijs/pixijs/assets/4526795/26c88b78-5352-4999-a3d6-ab3a8008b7a9)

The warning I get on the sprite test is: `[.WebGL-000014FC057B3F00] GL_INVALID_OPERATION: Invalid compressed image size.`
**Question 1:** Any idea why it throws this warning and shows black images?

I have also created [a npm package](https://www.npmjs.com/package/pixi-basis-ktx2) myself to load KTX2 with Pixi. I got it working in Pixi v7 and have no warnings with any of my tests. I have migrated this project to Pixi v8 and changed the loader to the Pixi KTX2 loader.

![Screenshot 2024-02-16 112724](https://github.com/pixijs/pixijs/assets/4526795/f09578c7-662e-44d7-968f-ec91fbdce009)

**Question 2:** Is there a reason we have to add the loadKTX and resolveCompressedTextureUrl to the correct parsers ourselves instead of being added standard to the extension? I know I have to do it for my package, because it's a separate package.

![Screenshot 2024-02-16 113001](https://github.com/pixijs/pixijs/assets/4526795/7b4201a4-7a2b-45a9-9d95-a7dfccd70410)

In Assets.ts:
![Screenshot 2024-02-16 113036](https://github.com/pixijs/pixijs/assets/4526795/6a306816-b120-42ba-84bd-9eb8d57d12a4)

Thanks in advance!

**Comment by Sparcks** [comment]
**Update on Question 1:**
It seems the warning and black images are only an issue on Desktop WebGL. It seems to be working on Android and iOS for WebGL & WebGPU, as well as on Desktop WebGPU.

Desktop (WebGPU):
![Screenshot 2024-02-19 103621](https://github.com/pixijs/pixijs/assets/4526795/b0ad280e-701f-4666-8e29-23d987a2b789)
Android (Both):
![Screenshot_2024-02-19-10-17-41-43_40deb401b9ffe8e1df2f1cc5ba480b12](https://github.com/pixijs/pixijs/assets/4526795/fd1a9a8d-4eb8-4114-b097-93dedbb25f0e)
![Screenshot_2024-02-19-10-40-43-17_40deb401b9ffe8e1df2f1cc5ba480b12](https://github.com/pixijs/pixijs/assets/4526795/da26e45e-f187-451f-ab3b-b42f7fc003d8)
iPhone (WebGL):
![IMG_0163](https://github.com/pixijs/pixijs/assets/4526795/405c1905-e58c-4fde-9677-d9292d74b1e3)

**Comment by ddenisyuk** [comment, marked as answer]
I can assume that your basis textures are being converted to BC7 format on the desktop. It's important to note that for BC7 textures, when used in WebGL, the width and height must be multiples of 4, according to https://registry.khronos.org/webgl/extensions/EXT_texture_compression_bptc/

Try to change the size of your textures from 777x1198 to, lets say, 780x1200.

**Reply by Sparcks** [reply]
Ah, Didn't think about checking the exact BC format. My package only did BC3, so that will be the reason it worked for my package. Thank you for your reply and sorry for my late reply. I had computer issues and had to reinstall windows and all my apps again. I will try to update the texture size as soon as I have everything working again.

**Reply by Sparcks** [reply]
Jup, that did the trick indeed. Thanks man!
![afbeelding](https://github.com/pixijs/pixijs/assets/4526795/9da5deb8-9365-4040-8c09-5284802f8d79)

**Reply by ape-casear** [reply]
> Jup, that did the trick indeed. Thanks man! ![afbeelding](https://private-user-images.githubusercontent.com/4526795/307772016-9da5deb8-9365-4040-8c09-5284802f8d79.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDk2MTM3MDEsIm5iZiI6MTc0OTYxMzQwMSwicGF0aCI6Ii80NTI2Nzk1LzMwNzc3MjAxNi05ZGE1ZGViOC05MzY1LTQwNDAtOGMwOS01Mjg0ODAyZjhkNzkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MTFUMDM0MzIxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODlmZmQyMWQyZGI3YWI5ZmZmMzRlYTFhOWJiM2QzYTVjOGFiYWQyNTliOGY2MTViNzNhMThhZDNlNmIwMzFiNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.r1srCxgKO5VgYXDyhEXAdySGVuuJHLrEJnnfPIKuFCY)

hello friend.   the black problem still exists after I chose bc3 as transcode format.  can you share your method  how your package transcoded it to BC3?  I use basis universal to generate a ktx2 size 65*66.  then I change the source code in pixi.js to set bc3 at prior place.  then I use pixi.js v8.9.2  to load it.   through debug I sure the format is bc3.  but the error still exists “[.WebGL-0x5b5c02fd6300] GL_INVALID_OPERATION: glCompressedTexImage2DRobustANGLE: Invalid compressed image size.”

```
   const preferredTranscodedFormat = [
        'astc-4x4-unorm',
        'bc3-rgba-unorm',
        'bc7-rgba-unorm',
        'rgba8unorm',
    ]
```
```
alphaMode: "no-premultiply-alpha"
antialias: false
autoGarbageCollect: false
autoGenerateMipmaps: false
destroyed: false
dimension: "2d"
format: "bc3-rgba-unorm"
height: 66
isPowerOfTwo: false
```

**Reply by ape-casear** [reply]
> > Jup, that did the trick indeed. Thanks man! ![afbeelding](https://private-user-images.githubusercontent.com/4526795/307772016-9da5deb8-9365-4040-8c09-5284802f8d79.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDk2MTM3MDEsIm5iZiI6MTc0OTYxMzQwMSwicGF0aCI6Ii80NTI2Nzk1LzMwNzc3MjAxNi05ZGE1ZGViOC05MzY1LTQwNDAtOGMwOS01Mjg0ODAyZjhkNzkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MTFUMDM0MzIxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODlmZmQyMWQyZGI3YWI5ZmZmMzRlYTFhOWJiM2QzYTVjOGFiYWQyNTliOGY2MTViNzNhMThhZDNlNmIwMzFiNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.r1srCxgKO5VgYXDyhEXAdySGVuuJHLrEJnnfPIKuFCY)
> 
> hello friend. the black problem still exists after I chose bc3 as transcode format. can you share your method how your package transcoded it to BC3? I use basis universal to generate a ktx2 size 65*66. then I change the source code in pixi.js to set bc3 at prior place. then I use pixi.js v8.9.2 to load it. through debug I sure the format is bc3. but the error still exists “[.WebGL-0x5b5c02fd6300] GL_INVALID_OPERATION: glCompressedTexImage2DRobustANGLE: Invalid compressed image size.”
> 
> ```
>    const preferredTranscodedFormat = [
>         'astc-4x4-unorm',
>         'bc3-rgba-unorm',
>         'bc7-rgba-unorm',
>         'rgba8unorm',
>     ]
> ```
> 
> ```
> alphaMode: "no-premultiply-alpha"
> antialias: false
> autoGarbageCollect: false
> autoGenerateMipmaps: false
> destroyed: false
> dimension: "2d"
> format: "bc3-rgba-unorm"
> height: 66
> isPowerOfTwo: false
> ```



fixed it;  by aligning with and height in WORKER_CODE in pixi.js
```js
async function load(url) {
        const BasisTexture = await getBasis()
        const basisTexture = await fetchBasisTexture(url, BasisTexture)
        const levelBuffers = createLevelBuffers(
            basisTexture,
            basisTranscoderFormat
        )
        let width = basisTexture.getImageWidth(0, 0)
        let height = basisTexture.getImageHeight(0, 0)
        var dstBlockWidth = 4
        var dstBlockHeight = 4
        let alignedWidth =
            Math.floor((width + dstBlockWidth - 1) / dstBlockWidth) *
            dstBlockWidth
        let alignedHeight =
            Math.floor((height + dstBlockHeight - 1) / dstBlockHeight) *
            dstBlockHeight
        return {
            width: alignedWidth,
            height: alignedHeight,
            format: basisTranscodedTextureFormat,
            resource: levelBuffers,
            alphaMode: 'no-premultiply-alpha',
        }
    }
```

**Comment by Sparcks** [comment]
So the black images issue was indeed because of BC7 that needed multiples of 4 for the width and height size. My package transcoded it to BC3. 
If someone else has an answer for question 2, I'm still all ears 😄 
`Question 2: Is there a reason we have to add the loadKTX and resolveCompressedTextureUrl to the correct parsers ourselves instead of being added standard to the extension? I know I have to do it for my package, because it's a separate package.`

**Reply by ddenisyuk** [reply]
I had the same question; it seems it's an issue in v8: https://github.com/pixijs/pixijs/issues/10234#issuecomment-1959041924

**Reply by Sparcks** [reply]
Aha alright 😄 I will keep an eye on that post then, thanks again.
